File: activelog/services/data-export/src/exporters/archive_exporter.py
"""
Archive Export (ZIP, tar.gz) with preserved directory structure and metadata
"""

import os
import logging
import zipfile
import tarfile
import json
from datetime import datetime
from pathlib import Path
from typing import Dict, List, Any, Callable, Optional
import shutil

from ..core.config import settings
from ..utils.file_utils import get_file_metadata, get_files_by_filters
from ..utils.metadata_extractor import MetadataExtractor

logger = logging.getLogger(__name__)

class ArchiveExporter:
    """Archive export supporting ZIP and tar.gz formats"""
    
    def __init__(self):
        self.metadata_extractor = MetadataExtractor()
        
    async def initialize(self):
        """Initialize archive exporter"""
        logger.info("Archive exporter initialized")
    
    def export(self, job_id: str, config: Dict[str, Any], 
              filters: Dict[str, Any], output_dir: str,
              progress_callback: Callable) -> Dict[str, Any]:
        """
        Export files to archive format (ZIP or tar.gz)
        
        Args:
            job_id: Export job identifier
            config: Archive export configuration
            filters: File selection filters
            output_dir: Output directory path
            progress_callback: Progress reporting callback
            
        Returns:
            Export result with file path and metadata
        """
        
        try:
            progress_callback(0, 100, "Initializing archive export")
            
            # Get files to export
            files = get_files_by_filters(filters)
            total_files = len(files)
            
            if total_files == 0:
                return {"success": False, "error": "No files found matching filters"}
            
            # Get archive configuration
            archive_config = self._get_archive_config(config)
            archive_format = archive_config["format"]
            
            # Create archive file
            if archive_format == "zip":
                output_filename = f"export_{job_id}.zip"
                result = self._create_zip_archive(
                    files, output_dir, output_filename, 
                    archive_config, progress_callback, total_files
                )
            else:  # tar.gz
                output_filename = f"export_{job_id}.tar.gz"
                result = self._create_tar_archive(
                    files, output_dir, output_filename,
                    archive_config, progress_callback, total_files
                )
            
            if result["success"]:
                output_path = os.path.join(output_dir, output_filename)
                progress_callback(100, 100, "Archive export completed")
                
                return {
                    "success": True,
                    "output_path": output_path,
                    "filename": output_filename,
                    "file_size": os.path.getsize(output_path),
                    "files_included": result.get("files_added", 0),
                    "compression_ratio": result.get("compression_ratio", 0)
                }
            else:
                return result
                
        except Exception as e:
            logger.error(f"Archive export failed for job {job_id}: {e}")
            return {"success": False, "error": str(e)}
    
    def _get_archive_config(self, config: Dict[str, Any]) -> Dict[str, Any]:
        """Get archive configuration with defaults"""
        
        return {
            "format": config.get("format", "zip"),  # zip or tar.gz
            "compression_level": config.get("compression_level", settings.export.compression_level),
            "preserve_structure": config.get("preserve_structure", True),
            "include_metadata": config.get("include_metadata", True),
            "include_hidden": config.get("include_hidden", False),
            "follow_symlinks": config.get("follow_symlinks", False),
            "exclude_patterns": config.get("exclude_patterns", []),
            "max_file_size": config.get("max_file_size", 0),  # 0 = no limit
            "create_manifest": config.get("create_manifest", True),
            "metadata_format": config.get("metadata_format", "json")  # json or txt
        }
    
    def _create_zip_archive(self, files: List[str], output_dir: str, 
                           filename: str, config: Dict[str, Any],
                           progress_callback: Callable, total_files: int) -> Dict[str, Any]:
        """Create ZIP archive"""
        
        try:
            output_path = os.path.join(output_dir, filename)
            
            # Compression method
            compression = zipfile.ZIP_DEFLATED
            compresslevel = config["compression_level"]
            
            original_size = 0
            compressed_size = 0
            files_added = 0
            
            with zipfile.ZipFile(output_path, 'w', compression=compression, 
                               compresslevel=compresslevel) as zipf:
                
                # Create manifest if requested
                manifest_data = []
                
                for i, file_path in enumerate(files):
                    try:
                        # Skip files that match exclude patterns
                        if self._should_exclude_file(file_path, config):
                            continue
                        
                        # Check file size limit
                        file_size = os.path.getsize(file_path)
                        if config["max_file_size"] > 0 and file_size > config["max_file_size"]:
                            logger.warning(f"Skipping large file {file_path}: {file_size} bytes")
                            continue
                        
                        # Determine archive path
                        if config["preserve_structure"]:
                            # Use relative path from common root
                            archive_path = self._get_relative_path(file_path, files)
                        else:
                            # Use just filename
                            archive_path = os.path.basename(file_path)
                        
                        # Add file to archive
                        zipf.write(file_path, archive_path)
                        
                        original_size += file_size
                        files_added += 1
                        
                        # Add to manifest
                        if config["create_manifest"]:
                            file_metadata = get_file_metadata(file_path)
                            manifest_data.append({
                                "original_path": file_path,
                                "archive_path": archive_path,
                                "size": file_size,
                                "modified": file_metadata.get("modified_time"),
                                "checksum": file_metadata.get("checksum")
                            })
                        
                        # Update progress
                        progress = 10 + (i / total_files * 80)
                        progress_callback(progress, 100, f"Added {files_added} files to ZIP")
                        
                    except Exception as e:
                        logger.warning(f"Failed to add file {file_path} to ZIP: {e}")
                
                # Add manifest file
                if config["create_manifest"] and manifest_data:
                    progress_callback(90, 100, "Creating manifest")
                    self._add_manifest_to_zip(zipf, manifest_data, config)
                
                # Add metadata files if requested
                if config["include_metadata"]:
                    progress_callback(95, 100, "Adding metadata files")
                    self._add_metadata_to_zip(zipf, files, config)
            
            # Calculate compression ratio
            compressed_size = os.path.getsize(output_path)
            compression_ratio = original_size / compressed_size if compressed_size > 0 else 1.0
            
            return {
                "success": True,
                "files_added": files_added,
                "original_size": original_size,
                "compressed_size": compressed_size,
                "compression_ratio": compression_ratio
            }
            
        except Exception as e:
            logger.error(f"Failed to create ZIP archive: {e}")
            return {"success": False, "error": str(e)}
    
    def _create_tar_archive(self, files: List[str], output_dir: str,
                           filename: str, config: Dict[str, Any], 
                           progress_callback: Callable, total_files: int) -> Dict[str, Any]:
        """Create tar.gz archive"""
        
        try:
            output_path = os.path.join(output_dir, filename)
            
            original_size = 0
            files_added = 0
            
            with tarfile.open(output_path, 'w:gz', compresslevel=config["compression_level"]) as tarf:
                
                # Create manifest if requested
                manifest_data = []
                
                for i, file_path in enumerate(files):
                    try:
                        # Skip files that match exclude patterns
                        if self._should_exclude_file(file_path, config):
                            continue
                        
                        # Check file size limit
                        file_size = os.path.getsize(file_path)
                        if config["max_file_size"] > 0 and file_size > config["max_file_size"]:
                            logger.warning(f"Skipping large file {file_path}: {file_size} bytes")
                            continue
                        
                        # Determine archive path
                        if config["preserve_structure"]:
                            archive_path = self._get_relative_path(file_path, files)
                        else:
                            archive_path = os.path.basename(file_path)
                        
                        # Add file to archive
                        tarf.add(file_path, arcname=archive_path, recursive=False)
                        
                        original_size += file_size
                        files_added += 1
                        
                        # Add to manifest
                        if config["create_manifest"]:
                            file_metadata = get_file_metadata(file_path)
                            manifest_data.append({
                                "original_path": file_path,
                                "archive_path": archive_path,
                                "size": file_size,
                                "modified": file_metadata.get("modified_time"),
                                "checksum": file_metadata.get("checksum")
                            })
                        
                        # Update progress
                        progress = 10 + (i / total_files * 80)
                        progress_callback(progress, 100, f"Added {files_added} files to tar.gz")
                        
                    except Exception as e:
                        logger.warning(f"Failed to add file {file_path} to tar.gz: {e}")
                
                # Add manifest file
                if config["create_manifest"] and manifest_data:
                    progress_callback(90, 100, "Creating manifest")
                    self._add_manifest_to_tar(tarf, manifest_data, config)
                
                # Add metadata files if requested
                if config["include_metadata"]:
                    progress_callback(95, 100, "Adding metadata files")
                    self._add_metadata_to_tar(tarf, files, config)
            
            # Calculate compression ratio
            compressed_size = os.path.getsize(output_path)
            compression_ratio = original_size / compressed_size if compressed_size > 0 else 1.0
            
            return {
                "success": True,
                "files_added": files_added,
                "original_size": original_size,
                "compressed_size": compressed_size,
                "compression_ratio": compression_ratio
            }
            
        except Exception as e:
            logger.error(f"Failed to create tar.gz archive: {e}")
            return {"success": False, "error": str(e)}
    
    def _should_exclude_file(self, file_path: str, config: Dict[str, Any]) -> bool:
        """Check if file should be excluded"""
        
        # Check hidden files
        if not config["include_hidden"]:
            if os.path.basename(file_path).startswith('.'):
                return True
        
        # Check exclude patterns
        for pattern in config.get("exclude_patterns", []):
            if pattern in file_path:
                return True
        
        return False
    
    def _get_relative_path(self, file_path: str, all_files: List[str]) -> str:
        """Get relative path for archive maintaining directory structure"""
        
        try:
            # Find common root directory
            common_path = os.path.commonpath(all_files)
            
            # Get relative path from common root
            relative_path = os.path.relpath(file_path, common_path)
            
            return relative_path
            
        except Exception:
            # Fallback to basename if can't determine relative path
            return os.path.basename(file_path)
    
    def _add_manifest_to_zip(self, zipf: zipfile.ZipFile, manifest_data: List[Dict], 
                            config: Dict[str, Any]):
        """Add manifest file to ZIP archive"""
        
        try:
            manifest = {
                "created": datetime.utcnow().isoformat(),
                "format": "zip",
                "total_files": len(manifest_data),
                "files": manifest_data
            }
            
            if config["metadata_format"] == "json":
                manifest_content = json.dumps(manifest, indent=2, default=str)
                manifest_filename = "MANIFEST.json"
            else:
                # Text format
                lines = [
                    f"Archive Manifest",
                    f"Created: {manifest['created']}",
                    f"Format: {manifest['format']}",
                    f"Total Files: {manifest['total_files']}",
                    "",
                    "Files:"
                ]
                
                for file_data in manifest_data:
                    lines.append(f"  {file_data['archive_path']}")
                    lines.append(f"    Original: {file_data['original_path']}")
                    lines.append(f"    Size: {file_data['size']} bytes")
                    if file_data.get('modified'):
                        lines.append(f"    Modified: {file_data['modified']}")
                    lines.append("")
                
                manifest_content = "\n".join(lines)
                manifest_filename = "MANIFEST.txt"
            
            # Add manifest to archive
            zipf.writestr(manifest_filename, manifest_content)
            
        except Exception as e:
            logger.warning(f"Failed to create manifest: {e}")
    
    def _add_manifest_to_tar(self, tarf: tarfile.TarFile, manifest_data: List[Dict],
                            config: Dict[str, Any]):
        """Add manifest file to tar.gz archive"""
        
        try:
            manifest = {
                "created": datetime.utcnow().isoformat(),
                "format": "tar.gz",
                "total_files": len(manifest_data),
                "files": manifest_data
            }
            
            if config["metadata_format"] == "json":
                manifest_content = json.dumps(manifest, indent=2, default=str)
                manifest_filename = "MANIFEST.json"
            else:
                # Text format
                lines = [
                    f"Archive Manifest",
                    f"Created: {manifest['created']}",
                    f"Format: {manifest['format']}",
                    f"Total Files: {manifest['total_files']}",
                    "",
                    "Files:"
                ]
                
                for file_data in manifest_data:
                    lines.append(f"  {file_data['archive_path']}")
                    lines.append(f"    Original: {file_data['original_path']}")
                    lines.append(f"    Size: {file_data['size']} bytes")
                    if file_data.get('modified'):
                        lines.append(f"    Modified: {file_data['modified']}")
                    lines.append("")
                
                manifest_content = "\n".join(lines)
                manifest_filename = "MANIFEST.txt"
            
            # Create tarinfo and add to archive
            tarinfo = tarfile.TarInfo(name=manifest_filename)
            tarinfo.size = len(manifest_content.encode('utf-8'))
            tarinfo.mtime = datetime.utcnow().timestamp()
            
            from io import BytesIO
            tarf.addfile(tarinfo, BytesIO(manifest_content.encode('utf-8')))
            
        except Exception as e:
            logger.warning(f"Failed to create manifest: {e}")
    
    def _add_metadata_to_zip(self, zipf: zipfile.ZipFile, files: List[str], 
                            config: Dict[str, Any]):
        """Add metadata files to ZIP archive"""
        
        try:
            # Create comprehensive metadata file
            metadata_collection = {
                "export_info": {
                    "created": datetime.utcnow().isoformat(),
                    "total_files": len(files),
                    "format": "zip"
                },
                "files": []
            }
            
            # Extract metadata for each file
            for file_path in files[:100]:  # Limit to first 100 files for metadata
                try:
                    file_metadata = self.metadata_extractor.extract_all_metadata(file_path)
                    
                    metadata_collection["files"].append({
                        "path": file_path,
                        "metadata": file_metadata
                    })
                    
                except Exception as e:
                    logger.warning(f"Failed to extract metadata for {file_path}: {e}")
            
            # Add metadata file
            if config["metadata_format"] == "json":
                metadata_content = json.dumps(metadata_collection, indent=2, default=str)
                zipf.writestr("metadata.json", metadata_content)
            else:
                # Create text format metadata
                lines = []
                lines.append("File Metadata Export")
                lines.append(f"Created: {metadata_collection['export_info']['created']}")
                lines.append(f"Total Files: {metadata_collection['export_info']['total_files']}")
                lines.append("")
                
                for file_data in metadata_collection["files"]:
                    lines.append(f"File: {file_data['path']}")
                    for key, value in file_data["metadata"].items():
                        lines.append(f"  {key}: {value}")
                    lines.append("")
                
                metadata_content = "\n".join(lines)
                zipf.writestr("metadata.txt", metadata_content)
            
        except Exception as e:
            logger.warning(f"Failed to add metadata files: {e}")
    
    def _add_metadata_to_tar(self, tarf: tarfile.TarFile, files: List[str],
                            config: Dict[str, Any]):
        """Add metadata files to tar.gz archive"""
        
        try:
            # Create comprehensive metadata file
            metadata_collection = {
                "export_info": {
                    "created": datetime.utcnow().isoformat(),
                    "total_files": len(files),
                    "format": "tar.gz"
                },
                "files": []
            }
            
            # Extract metadata for each file
            for file_path in files[:100]:  # Limit to first 100 files for metadata
                try:
                    file_metadata = self.metadata_extractor.extract_all_metadata(file_path)
                    
                    metadata_collection["files"].append({
                        "path": file_path,
                        "metadata": file_metadata
                    })
                    
                except Exception as e:
                    logger.warning(f"Failed to extract metadata for {file_path}: {e}")
            
            # Add metadata file
            if config["metadata_format"] == "json":
                metadata_content = json.dumps(metadata_collection, indent=2, default=str)
                filename = "metadata.json"
            else:
                # Create text format metadata
                lines = []
                lines.append("File Metadata Export")
                lines.append(f"Created: {metadata_collection['export_info']['created']}")
                lines.append(f"Total Files: {metadata_collection['export_info']['total_files']}")
                lines.append("")
                
                for file_data in metadata_collection["files"]:
                    lines.append(f"File: {file_data['path']}")
                    for key, value in file_data["metadata"].items():
                        lines.append(f"  {key}: {value}")
                    lines.append("")
                
                metadata_content = "\n".join(lines)
                filename = "metadata.txt"
            
            # Create tarinfo and add to archive
            tarinfo = tarfile.TarInfo(name=filename)
            tarinfo.size = len(metadata_content.encode('utf-8'))
            tarinfo.mtime = datetime.utcnow().timestamp()
            
            from io import BytesIO
            tarf.addfile(tarinfo, BytesIO(metadata_content.encode('utf-8')))
            
        except Exception as e:
            logger.warning(f"Failed to add metadata files: {e}")
    
    def extract_archive(self, archive_path: str, extract_to: str) -> Dict[str, Any]:
        """Extract archive (utility method)"""
        
        try:
            # Determine archive type
            if archive_path.endswith('.zip'):
                return self._extract_zip(archive_path, extract_to)
            elif archive_path.endswith(('.tar.gz', '.tgz')):
                return self._extract_tar(archive_path, extract_to)
            else:
                return {"success": False, "error": "Unsupported archive format"}
                
        except Exception as e:
            logger.error(f"Failed to extract archive {archive_path}: {e}")
            return {"success": False, "error": str(e)}
    
    def _extract_zip(self, zip_path: str, extract_to: str) -> Dict[str, Any]:
        """Extract ZIP archive"""
        
        try:
            os.makedirs(extract_to, exist_ok=True)
            
            with zipfile.ZipFile(zip_path, 'r') as zipf:
                zipf.extractall(extract_to)
                file_count = len(zipf.namelist())
            
            return {
                "success": True,
                "extracted_files": file_count,
                "extract_path": extract_to
            }
            
        except Exception as e:
            return {"success": False, "error": str(e)}
    
    def _extract_tar(self, tar_path: str, extract_to: str) -> Dict[str, Any]:
        """Extract tar.gz archive"""
        
        try:
            os.makedirs(extract_to, exist_ok=True)
            
            with tarfile.open(tar_path, 'r:gz') as tarf:
                tarf.extractall(extract_to)
                file_count = len(tarf.getnames())
            
            return {
                "success": True,
                "extracted_files": file_count,
                "extract_path": extract_to
            }
            
        except Exception as e:
            return {"success": False, "error": str(e)}

File: activelog/services/data-export/src/exporters/archiveExporter.js
const archiver = require('archiver');
const tar = require('tar');
const fs = require('fs-extra');
const path = require('path');
const crypto = require('crypto');
const logger = require('../utils/logger');

class ArchiveExporter {
  constructor(options = {}) {
    this.options = {
      compressionLevel: options.compressionLevel || 6,
      includeMetadata: options.includeMetadata !== false,
      preservePermissions: options.preservePermissions !== false,
      generateChecksum: options.generateChecksum !== false,
      excludePatterns: options.excludePatterns || [],
      maxFileSize: options.maxFileSize || 1024 * 1024 * 1024, // 1GB
      ...options
    };
  }

  async createZipArchive(sources, outputPath, metadata = {}) {
    try {
      const output = fs.createWriteStream(outputPath);
      const archive = archiver('zip', {
        zlib: { level: this.options.compressionLevel }
      });

      archive.pipe(output);

      if (this.options.includeMetadata) {
        const metadataContent = this.generateMetadataFile(metadata, sources);
        archive.append(metadataContent, { name: 'export_metadata.json' });
      }

      await this.addSourcesToArchive(archive, sources);

      await archive.finalize();

      return new Promise((resolve, reject) => {
        output.on('close', async () => {
          const stats = fs.statSync(outputPath);
          const result = {
            success: true,
            outputPath,
            fileSize: stats.size,
            totalBytes: archive.pointer(),
            compressionRatio: (1 - stats.size / archive.pointer()) * 100
          };

          if (this.options.generateChecksum) {
            result.checksum = await this.generateChecksum(outputPath);
          }

          logger.info(`ZIP archive created: ${outputPath} (${this.formatBytes(stats.size)})`);
          resolve(result);
        });

        output.on('error', reject);
        archive.on('error', reject);
      });
    } catch (error) {
      logger.error('ZIP archive creation failed:', error);
      throw error;
    }
  }

  async createTarGzArchive(sources, outputPath, metadata = {}) {
    try {
      const tempDir = path.join(path.dirname(outputPath), 'temp_tar_' + Date.now());
      await fs.ensureDir(tempDir);

      const archiveDir = path.join(tempDir, 'export');
      await fs.ensureDir(archiveDir);

      if (this.options.includeMetadata) {
        const metadataContent = this.generateMetadataFile(metadata, sources);
        await fs.writeFile(
          path.join(archiveDir, 'export_metadata.json'),
          metadataContent
        );
      }

      await this.copySourcesToTempDir(sources, archiveDir);

      await tar.create(
        {
          gzip: true,
          file: outputPath,
          cwd: tempDir,
          portable: true,
          preservePaths: false
        },
        ['export']
      );

      await fs.remove(tempDir);

      const stats = fs.statSync(outputPath);
      const result = {
        success: true,
        outputPath,
        fileSize: stats.size
      };

      if (this.options.generateChecksum) {
        result.checksum = await this.generateChecksum(outputPath);
      }

      logger.info(`TAR.GZ archive created: ${outputPath} (${this.formatBytes(stats.size)})`);
      return result;
    } catch (error) {
      logger.error('TAR.GZ archive creation failed:', error);
      throw error;
    }
  }

  async addSourcesToArchive(archive, sources) {
    for (const source of sources) {
      if (this.shouldExclude(source.path)) {
        continue;
      }

      const stats = await fs.stat(source.path);

      if (stats.size > this.options.maxFileSize) {
        logger.warn(`Skipping large file: ${source.path} (${this.formatBytes(stats.size)})`);
        continue;
      }

      if (stats.isDirectory()) {
        archive.directory(source.path, source.name || path.basename(source.path));
      } else {
        const options = {
          name: source.name || path.basename(source.path)
        };

        if (this.options.preservePermissions) {
          options.mode = stats.mode;
        }

        if (source.metadata) {
          options.comment = JSON.stringify(source.metadata);
        }

        archive.file(source.path, options);
      }
    }
  }

  async copySourcesToTempDir(sources, targetDir) {
    for (const source of sources) {
      if (this.shouldExclude(source.path)) {
        continue;
      }

      const targetPath = path.join(targetDir, source.name || path.basename(source.path));
      
      const stats = await fs.stat(source.path);
      if (stats.size > this.options.maxFileSize) {
        logger.warn(`Skipping large file: ${source.path}`);
        continue;
      }

      if (stats.isDirectory()) {
        await fs.copy(source.path, targetPath, {
          preserveTimestamps: true,
          filter: (src) => !this.shouldExclude(src)
        });
      } else {
        await fs.copy(source.path, targetPath, {
          preserveTimestamps: true
        });
      }

      if (source.metadata && this.options.includeMetadata) {
        await fs.writeFile(
          targetPath + '.metadata.json',
          JSON.stringify(source.metadata, null, 2)
        );
      }
    }
  }

  generateMetadataFile(metadata, sources) {
    const exportMetadata = {
      exportDate: new Date().toISOString(),
      exportVersion: '1.0.0',
      sources: sources.map(source => ({
        originalPath: source.path,
        archivePath: source.name || path.basename(source.path),
        metadata: source.metadata || {}
      })),
      options: {
        compressionLevel: this.options.compressionLevel,
        preservePermissions: this.options.preservePermissions
      },
      ...metadata
    };

    return JSON.stringify(exportMetadata, null, 2);
  }

  shouldExclude(filePath) {
    for (const pattern of this.options.excludePatterns) {
      if (typeof pattern === 'string') {
        if (filePath.includes(pattern)) {
          return true;
        }
      } else if (pattern instanceof RegExp) {
        if (pattern.test(filePath)) {
          return true;
        }
      }
    }
    return false;
  }

  async generateChecksum(filePath, algorithm = 'sha256') {
    return new Promise((resolve, reject) => {
      const hash = crypto.createHash(algorithm);
      const stream = fs.createReadStream(filePath);

      stream.on('data', data => hash.update(data));
      stream.on('end', () => resolve(hash.digest('hex')));
      stream.on('error', reject);
    });
  }

  formatBytes(bytes) {
    const sizes = ['Bytes', 'KB', 'MB', 'GB', 'TB'];
    if (bytes === 0) return '0 Bytes';
    const i = Math.floor(Math.log(bytes) / Math.log(1024));
    return Math.round(bytes / Math.pow(1024, i) * 100) / 100 + ' ' + sizes[i];
  }

  async extractArchive(archivePath, extractDir, options = {}) {
    try {
      await fs.ensureDir(extractDir);

      const ext = path.extname(archivePath).toLowerCase();

      if (ext === '.zip') {
        return await this.extractZip(archivePath, extractDir, options);
      } else if (ext === '.gz' || ext === '.tar') {
        return await this.extractTarGz(archivePath, extractDir, options);
      } else {
        throw new Error(`Unsupported archive format: ${ext}`);
      }
    } catch (error) {
      logger.error('Archive extraction failed:', error);
      throw error;
    }
  }

  async extractZip(zipPath, extractDir, options) {
    const yauzl = require('yauzl');
    
    return new Promise((resolve, reject) => {
      yauzl.open(zipPath, { lazyEntries: true }, (err, zipfile) => {
        if (err) return reject(err);

        const extractedFiles = [];

        zipfile.readEntry();
        zipfile.on('entry', (entry) => {
          const entryPath = path.join(extractDir, entry.fileName);
          
          if (/\/$/.test(entry.fileName)) {
            fs.ensureDir(entryPath).then(() => zipfile.readEntry());
          } else {
            zipfile.openReadStream(entry, (err, readStream) => {
              if (err) return reject(err);

              fs.ensureDir(path.dirname(entryPath)).then(() => {
                const writeStream = fs.createWriteStream(entryPath);
                readStream.pipe(writeStream);
                
                writeStream.on('close', () => {
                  extractedFiles.push(entryPath);
                  zipfile.readEntry();
                });
              });
            });
          }
        });

        zipfile.on('end', () => {
          resolve({
            success: true,
            extractedFiles,
            extractDir
          });
        });
      });
    });
  }

  async extractTarGz(tarPath, extractDir, options) {
    await tar.extract({
      file: tarPath,
      cwd: extractDir,
      strict: true
    });

    const extractedFiles = await this.getExtractedFiles(extractDir);

    return {
      success: true,
      extractedFiles,
      extractDir
    };
  }

  async getExtractedFiles(dir, files = []) {
    const items = await fs.readdir(dir);
    
    for (const item of items) {
      const fullPath = path.join(dir, item);
      const stats = await fs.stat(fullPath);
      
      if (stats.isDirectory()) {
        await this.getExtractedFiles(fullPath, files);
      } else {
        files.push(fullPath);
      }
    }
    
    return files;
  }

  async createSplitArchive(sources, outputPath, maxSplitSize = 100 * 1024 * 1024) { // 100MB
    const tempDir = path.join(path.dirname(outputPath), 'temp_split_' + Date.now());
    await fs.ensureDir(tempDir);

    const splits = [];
    let currentSplit = [];
    let currentSize = 0;

    for (const source of sources) {
      const stats = await fs.stat(source.path);
      
      if (currentSize + stats.size > maxSplitSize && currentSplit.length > 0) {
        splits.push([...currentSplit]);
        currentSplit = [];
        currentSize = 0;
      }
      
      currentSplit.push(source);
      currentSize += stats.size;
    }

    if (currentSplit.length > 0) {
      splits.push(currentSplit);
    }

    const splitFiles = [];
    for (let i = 0; i < splits.length; i++) {
      const splitPath = outputPath.replace(/\.zip$/, `.part${i + 1}.zip`);
      const result = await this.createZipArchive(splits[i], splitPath);
      splitFiles.push(result);
    }

    await fs.remove(tempDir);

    return {
      success: true,
      splitFiles,
      totalParts: splits.length
    };
  }
}

module.exports = ArchiveExporter;